qip-0012.md

```diff
@@ -0,0 +1,51 @@
+```
+ QIP: 12
+ Layer: Consensus (hard fork)
+ Title: Qi UTXO Pruning
+ Author: wizeguyy <wizeguyy+qip@quai.org>
+ Comments-Summary: No comments yet.
+ Comments-URI: https://github.com/quainetwork/qips/wiki/Comments:QIP-0012
+ Status: Draft
+ Type: Standards Track
+ Created: 2024-04-02
+ License: BSD-2-Clause
+```
+
+## Abstract
+This QIP proposes a mechanism to prune dust UTXOs from the ledger.
+
+## Motivation
+Transaction outputs in the Qi UTXO ledger, occupy space and impose a burden on
+UTXO transaction processing, from the set root recomputations. This is not a
+problem under normal usage, but can become a problem if any users/wallets are
+lazy or irresponsible with UTXO reconsolidation, or worse if the network
+undergoes a spam attack. This QIP describes a protocol change to disincentivize
+UTXO state pollution without meaningfully impacting normal transactional
+behavior.
+
+## Rationale
+Normal Qi wallet behavior will involve some amount of UTXO splitting, to create
+change outputs for payments, as well as UTXO reconsolidation as the wallet
+acquires too many small change outputs. This QIP should not negatively impact
+this expected user behavior, but should provide a cost to lazy wallets that do
+not reconsolidate change outputs.
+
+## Specification
+We achieve this simply, by limiting the size of the UTXO set. If a transaction
+creates new UTXOs which would exceed the UTXO set capacity, we destroy the
+smallest UTXOs in the ledger. This is practical to do thanks to the fixed
+denominations in the Qi ledger.
```

Comment:
How do I find the smallest and oldest UTXO in the ledger? The UTXO trie is not organized in a FIFO manner (or any organization except for some key prefixing, as far as I can tell)

Reply:
Yep, it will involve some filter routine on these events. As an optimization, we can consider indexing by denomination or something like that

Reply:
Can you add some detail in the QIP regarding how this might be achieved? I was under the impression that indexing would be optional, not in consensus

Reply:
well this is a different kind of indexing than the indexers used by the RPC. It _could_ be implemented any number of ways, up to each implementation, so I don't want the QIP to say "this is the way its done". But for context, here are some ways it could be done:

## Just-In-Time Scanning (not performant):
```rust
let mut denomination = MAX_DENOMINATION;
let mut delete_list = Vec::new();

// First scan and collect the keys of every UTXO to be deleted
for utxo in set {
    // Found a new smaller denomination. Reset the scanner.
    if utxo.denomination < denomination {
        denomination = utxo.denomination;
        delete_list.clear();
    }
    
    // If the utxo matches the smallest denomination, add it to the delete list
    if utxo.denomination == denomination {
        delete_list.push(utxo.key);
    }
}

// Now go back and delete each key you found
for key in delete_list {
    set.delete(key);
}
```

## Keeping Denominations By Index:
```rust
struct UtxoSet {
    utxos: HashMap<UtxoKey, Utxo>,
    denominations: HashMap<Denomination, HashSet<UtxoKey>>,
}
impl UtxoSet {
    // Add a UTXO to the set, and prune the set if it gets too large
    fn AddUtxo(mut self, utxo: Utxo) {
        ... make sure its a valid utxo ...

        // Add to the UTXO set
        self.utxos.insert(utxo.key, utxo);

        // Add to the denomination index
        self.denominations[utxo.Denomination].insert(utxo.key)

        // Check if the set is too large, and trigger deletions
        if self.len() > UTXO_SET_CAPACITY {
            // Find the smallest denomination in the set
            let min_denomination = self.denominations
                .iter()                                 // iterate through the index lists
                .filter(|(den, list)| !list.is_empty()) // filter out any denominations which don't have existing UTXOs
                .map(|(den, _)| den)                    // just look at the denominations
                .min();                                 // get the smallest denomination
            
            // Delete every UTXO in the smallest denomination list
            for key in self.denominations.get(min_denomination) {
                self.DeleteUtxo(key);
            }
        }
    }

    // Delete a UTXO from the set
    fn DeleteUtxo(mut self, key: UtxoKey) {
        // Delete it from the set, and if it existed, delte it from the indexed lists
        if let Some(utxo) = self.utxos.remove(key) {
            self.denominations.get(utxo.denomination).remove(utxo.key);
        }
    }
}
```

The second requires more memory (effectively double the UTXO set), but takes very little time to prune the set.

There could be trade-off approaches, e.g. one which only indexes the keys of the smallest denomination, but the logic to get that right is beyond the scope of this thread, lol

Reply:
The deletion should also take into account when the UTXO was created, right? i.e. the smallest *and* oldest, dustiest UTXOs are deleted first

Reply:
Perhaps an ordered list should be maintained, organized by FIFO and denomination. It would have to be updated for each block, and perhaps even committed to in the header. Hopefully insertions are no worse than O(logn)...

Reply:
yeah, I just gave some quick examples. There's a million ways to skin this cat